server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +52,267 @@
 
+    return topCandidate;
+  }
 
-    return stats;
+  private static void assignReplicas(
+      final Object2IntMap<String> targetReplicants,
+      final Object2IntMap<String> currentReplicants,
+      final DruidCoordinatorRuntimeParams params,
+      final DataSegment segment,
+      final CoordinatorStats stats
+  )
+  {
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final String tier = entry.getKey();
+      final int numAssigned = assignReplicasForTier(
+          tier,
+          entry.getIntValue(),
+          currentReplicants.getOrDefault(tier, 0),
+          params,
+          getHolderList(tier, params.getDruidCluster(), createPredicate(params)),
+          segment
+      );
+      stats.addToTieredStat(ASSIGNED_COUNT, tier, numAssigned);
+    }
   }
 
-  private int assign(
-      final ReplicationThrottler replicationManager,
+  private static int assignReplicasForTier(
       final String tier,
-      final int totalReplicantsInCluster,
-      final int expectedReplicantsInTier,
-      final int totalReplicantsInTier,
-      final BalancerStrategy strategy,
-      final List<ServerHolder> serverHolderList,
+      final int targetReplicantsInTier,
+      final int currentReplicantsInTier,
+      final DruidCoordinatorRuntimeParams params,
+      final List<ServerHolder> holders,
       final DataSegment segment
   )
   {
-    int assignedCount = 0;
-    int currReplicantsInTier = totalReplicantsInTier;
-    int currTotalReplicantsInCluster = totalReplicantsInCluster;
-    while (currReplicantsInTier < expectedReplicantsInTier) {
-      boolean replicate = currTotalReplicantsInCluster > 0;
+    final int numToAssign = targetReplicantsInTier - currentReplicantsInTier;
+    if (numToAssign <= 0 || holders.isEmpty()) {
+      return 0;
+    }
 
-      if (replicate && !replicationManager.canCreateReplicant(tier)) {
-        break;
+    final ReplicationThrottler throttler = params.getReplicationManager();
+    for (int numAssigned = 0; numAssigned < numToAssign; numAssigned++) {
+      if (!throttler.canCreateReplicant(tier)) {
+        return numAssigned;
       }
 
-      final ServerHolder holder = strategy.findNewSegmentHomeReplicator(segment, serverHolderList);
-
+      final ServerHolder holder = params.getBalancerStrategy().findNewSegmentHomeReplicator(segment, holders);
       if (holder == null) {
         log.warn(
             "Not enough [%s] servers or node capacity to assign segment[%s]! Expected Replicants[%d]",
-            tier,
-            segment.getIdentifier(),
-            expectedReplicantsInTier
-        );
-        break;
-      }
-
-      if (replicate) {
-        replicationManager.registerReplicantCreation(
-            tier, segment.getIdentifier(), holder.getServer().getHost()
+            tier, segment.getIdentifier(), targetReplicantsInTier
         );
+        return numAssigned;
       }
+      holders.remove(holder);
```

Comment:
Why added this?

Reply:
This is to prevent the holder from serving more than one replica of the same segment. During the current run, I believe the assignment will not be immediately reflected in the `ServerHolder`. Have I mistaken on the expected behavior?